### ISSUE TITLE
Parse Server may expose HTTP routes before complete initialization

### ISSUE THREAD

**Issue by sadortun** [parse-community/parse-server#7527]
### New Issue Checklist
- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).
- [x] I can reproduce the issue with the [latest version of Parse Server](https://github.com/parse-community/parse-server/releases). 

### Issue Description
Currently, all `ParseServer` initialization is done in the `constructor` and there is no way to ensure everything is ready before the app is loaded.

#### The main issue(s)

- Parse Server constructor will continue to do initialization tasks long after the constructor caller returns, and we have no way to await for its completion.

- Server will start to listen to incoming http requests before CloudCode is loaded.

--- 

This can be bad in some case because `express` and `ParseServer` can get exposed to the world, before it have completed its initialization leading to client side errors 

Ideally, we should refactor the `constructor` startup process into a `async start()` method that can be awaited on.

Currently we see at occasion during rolling updates clients running cloud functions `Parse.Cloud.run('xyz')` and receiving `Parse.Error`  `Function XYZ does not exists`. This is caused by having the parse `api` mounted and `express` running before all Cloud Code have a time to start.

# A few solutions

## (A) Without introducing a BC Break

This solution is great as a compromise for now..  It's simple, and have no long term implications.

Add a `isReady()`  function that return a promise that is resolved when the constructor has completed
```js
class ParseServer {
  private isReady : Promise<void>
   constructor(){
       this.isReady = new Promise<void>(resolve => {
          // ... ALL CURRENT CONSTRUCTOR CODE
       })       
   }
   
   public function isReady () : Promise<void> {
      return this.isReady;
   }
}
```

```js
// ......  server.ts
const api = new ParseServer(config)
await api.ready()
```

## (B) Without introducing a BC Break, and a path forward

This solution is great long term solution.

Move constructor init into a `start()` method, and add a `ParseServerOption`  `autoStart: true`.  Later in version `6.0.0` we can default `autoStart` to `false` or ignore/deprecate the parameter
```js
class ParseServer {
   constructor(options){
      if(options.autoStart === true){
          this.start()
      }
   }
   
   public function start() : Promise<void> {
       // ... ALL CURRENT CONSTRUCTOR CODE ...
   }
}
```

```js
// ......  server.ts
const config = { autoStart : false };
const api = new ParseServer(config)
await api.start()
```

## (C) Pure and simple BC Break

Just put all constructor code into a `start()` and tell users they need  to call it in the `5.0.0` release notes
```js
const api = new ParseServer(config)
await api.start()
```

--- 

(C) is obvisously a bit harsh 😆 !    LMK what you think and i can submit a PR for either (A) or (B)

### Failing Test Case / Pull Request

- [ ] 🤩 I submitted a PR with a fix and a test case.
- [ ] 🧐 I submitted a PR with a failing test case.

### Environment

Server
- Parse Server version: 4.5.0
- Operating system: linux/windows
- Local or remote host (AWS, Azure, Google Cloud, Heroku, Digital Ocean, etc): local a d GC

Database
- System (MongoDB or Postgres): `mingo`
- Database version: `recent`
- Local or remote host (MongoDB Atlas, mLab, AWS, Azure, Google Cloud, etc): `gc`

Client
- SDK (iOS, Android, JavaScript, PHP, Unity, etc): `js`
- SDK version: `4.5.4`

### Logs

Have a good day,
Samuel

**Comment by sadortun** [comment]
Hi @davimacedo,

Thanks for your suggestion! Yes, that would solve part of the issue, but Cloud Codes will still load after the `serverStartComplete` is triggered.

In general, performing `async` operations in a class constructor is a bad idea. My above suggestion `(B)` would address that, and ensure all proper `async` code can be awaited in a proper way. 

---

For my `(A)` suggestion, would be a minimal impact of you don't intend in having a distinct `start()` method. At first I thought it was just a workaround, but it seems it's [a documented design pattern](https://qwtel.com/posts/software/async-constructor-pattern/)

**Comment by mtrezza** [comment]
> Currently we see at occasion during rolling updates clients running cloud functions Parse.Cloud.run('xyz') and receiving Parse.Error Function XYZ does not exists.

Reading again the original issue, this sounds more like an infrastructure issue, because regardless of whether the express app listens to a port or not, you probably wouldn't want to send requests to the server instance in the first place. I can think of 2 design patterns that are usual in an infrastructure context:

- a) Parse Server calls an API to announce when it is has fully loaded all of its components and is ready to receive requests (push)
- b) Make an infrastructure component query a Parse Server API regularly to know when it is ready to receive requests (pull)

- For a) I assume that it would be `serverStartComplete` that can call an API.
- For b) I assume it would be the Parse Server `/status` endpoint that can be polled.

If `serverStartComplete` is called before everything is loaded or the `/status` endpoint indicates an incorrect readiness state, that would be the issue to address in my opinion.

**Comment by sadortun** [comment]
> Reading again the original issue, this sounds more like an infrastructure issue, because if the express app does not listen to the port yet

This is a ParseServer issue. Express starts to listen to requests before ParseServer finished it's loading.

Currently, it's not that of a big deal because overall, initialization is fairly quick 1-2 seconds. 

But in the case of Parse not being able to connect to Mongo, Parse and Express should not listen to requests before the server is ready OR report a healthy status when in reality any requests would fail due to the lack of database connection.

Secondly, this issue/improvement is directly related to the Schema Migration PR. With the introduction of schema migrations, you definitively DONT want Cloud code to load before migrations are completed, and you don't want Express to listen to requests before Cloud code is loaded.

Schema Migrations are quite fast, but if you have 50 Schemas to validate/update it may take a 10-15 seconds.

The fixes/improvements proposed above would fix theses issues.

**Comment by mtrezza** [comment]
> This is a ParseServer issue.

I agree if one of the following is true, like I said above:

> If `serverStartComplete` is called before everything is loaded or the `/status` endpoint indicates an incorrect readiness state, that would be the issue to address in my opinion.

I think you confirmed that already in your previous comment:

> Cloud Codes will still load after the serverStartComplete is triggered.

Does the construction need to change to fix that issue? I'm trying to break down this broad discussion (constructor change, existing callback not working, schema migration, etc) into smaller issues to see what can be addressed individually.

**Comment by sadortun** [comment]
@mtrezza 

> Does the constructor need to change to fix the callback issue?

Good question, IMO `serverStartCpmplete` should be called once everything is actually ready. Currently, it's not the case since it's called before ParseCloud.

That said, I have no idea if thats the intended behaviour or not.  

If you don't insight about this, I'll have a look in `git blame` to try to see how it got like it is now.

---

Side note for later: If we implement one of my suggestions above, i would also recommend deprecating in v6.0  `serverStartComplete` in favor or just having a straightforward `start()` or ,`isReady()` function we can `await` on. 

I'll open a different issue when this issue is resolved to address this later.

**Comment by mtrezza** [comment]
From what I can see, this is a bug.

The [docs](https://github.com/parse-community/parse-server/blob/9ea355b4635226ae4da17c8cc5fb0321e3fdec5e/src/Options/index.js#L242) say:

> Callback when server has started

The server has not started if it has not loaded all of its modules. As such, if people are using this callback as intended (as you would in your case like Davi suggested), then they will currently face the issue of errors because the callback is called too early.

Fixing this would not be considered a breaking change, because it is a bug fix.

My suggestion would be to fix this bug in this issue and treat the constructor change as an enhancement in a separate issue.

**Comment by sadortun** [comment]
> fix this bug in this issue 

Ok, awesome! I'll submit a PR sometime soon ...

> and treat the constructor change as an enhancement in a separate issue

Ok. Do we keep one issue with two PRs? Otherwise I'll duplicate this issue to discuss the next steps.

**Comment by mtrezza** [comment]
Yes, I suggest to split the enhancement into a separate issue and PR. This way we can fix the issue quicker and backport it to 4.x. It also gives us more flexibility for a breaking change with the constructor, should it be needed.

**Comment by SebC99** [comment]
I just discovered the issue.
In our case, we resolved this issue by creating a new route `/healthCheck` which resolves only when a boolean is set to true, which we do in `serverStartComplete`.
I guess the most surprising thing (which can be seen as a bug) is that the route `/parse/health` returns true even before `serverStartComplete` is called. 
So a "quick" way to fix would be to make sure the `/parse/health` route checks the completion status before sending OK.

This lets us config AWS load balancer to check a specific route to know the instance status.

**Comment by dblythy** [comment]
If we are supporting this as a breaking change for V6, we could also remove `serverStartComplete` and `startCallbackError` and use native async/await with `Error`.

```
  const api = new ParseServer(config);
  await api.start()
```

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.0.0-alpha.16](https://github.com/parse-community/parse-server/releases/tag/6.0.0-alpha.16)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.0.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/6.0.0-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.0.0](https://github.com/parse-community/parse-server/releases/tag/6.0.0)

**Comment by sadortun** [comment]
This is amazing work guys ! 

My deploy scripts will be so happy 🥂
